### ISSUE TITLE
Misuse of -Wl,-undefined

### ISSUE THREAD

**Issue by rui314** [KhronosGroup/glslang#3073]
`CMakeLists.txt` contains some code to pass `-undefined error` to the linker as shown below.

```
    if(NOT CMAKE_VERSION VERSION_LESS "3.13")
        # Error if there's symbols that are not found at link time.
        # add_link_options() was added in CMake 3.13 - if using an earlier
        # version don't set this - it should be caught by presubmits anyway.
        if (WIN32)
            add_link_options("-Wl,--no-undefined")
        else()
            add_link_options("-Wl,-undefined,error")
        endif()
    endif()
```

The problem is that `-undefined` isn't an option to force the linker to report an undefined symbol error. See https://man7.org/linux/man-pages/man1/ld.1.html. Did you confuse it with `--unresolved-symbols`?

**Comment by jeremy-lunarg** [comment]
That logic is for the clang compiler; and consequently, the LLVM linker. The man page you linked is for the GNU linker.

Please re-open if you disagree.

**Comment by rui314** [comment]
I disagree, but it looks like I don't have a privilege to reopen the issue.

Here is a quote from the LLVM linker's man page.

```
     --undefined=symbol, -u symbol
             If symbol is not defined after symbol resolution, and there's a
             static library that contains an object file defining the symbol,
             load the member to include the object file in the output file.
```

So `-undefined,error` is being accidentally interpreted as `-u "ndefined,errro"` which is obviously wrong.

The option that matches with the comment for this piece of code is `--unresolved-symbols=report-all`. Here is an quote from the GNU ld's man page.

```
       --unresolved-symbols=method
           Determine how to handle unresolved symbols.  There are four
           possible values for method:

           ignore-all
               Do not report any unresolved symbols.

           report-all
               Report all unresolved symbols.  This is the default.

           ignore-in-object-files
               Report unresolved symbols that are contained in shared
               libraries, but ignore them if they come from regular object
               files.

           ignore-in-shared-libs
               Report unresolved symbols that come from regular object files,
               but ignore them if they come from shared libraries.  This can
               be useful when creating a dynamic binary and it is known that
               all the shared libraries that it should be referencing are
               included on the linker's command line.

           The behaviour for shared libraries on their own can also be
           controlled by the --[no-]allow-shlib-undefined option.

           Normally the linker will generate an error message for each
           reported unresolved symbol but the option --warn-unresolved-symbols
           can change this to a warning.
```

**Comment by rui314** [comment]
Actually, `--no-undefined` should work both targets, so I'm not sure what `if(WIN32)` was intended for.

```
       --no-undefined
       -z defs
           Report unresolved symbol references from regular object files.
           This is done even if the linker is creating a non-symbolic shared
           library.  The switch --[no-]allow-shlib-undefined controls the
           behaviour for reporting unresolved references found in shared
           libraries being linked in.

           The effects of this option can be reverted by using "-z undefs".
```

**Comment by greg-lunarg** [comment]
The dominating condition for the original code above is that the compiler is clang. If using clang but not on WIN32, it is assumed that clang uses LLVM's lld, so add_link_options("-Wl,-undefined,error") is correct based on this reasonable assumption.

It is theoretically possible that someone would want to use GNU ld with clang, which we believe you are doing. Such a hybrid build is currently not supported by this script. If someone wishes to adapt this script to add this capability, we will consider it.

I will change this issue into a feature request for this capability.

**Comment by rui314** [comment]
No, I mean, `-Wl,-undefined,error` is just ambiguous because there's a single-letter `-u` option. It can either be taken as a `--undefined error` or `-u ndefined error` where `ndefined` is an argument for `-u` in the second case.

Just adding an extra hyphen like `-Wl,--undefined,error` disambiguates it.

**Comment by greg-lunarg** [comment]
After a little more research on the part of @jeremy-lunarg, the `-Wl,-undefined,error` syntax is not an attempt to use the GNU/LLVM ld -u option, but is a unique option from ld on Apple, so we are going to fix the code around that option to only be executed in that situation. I will restore the original title of this issue.

**Comment by jeremy-lunarg** [comment]
The GNU linker, `ld`, accepts `--no-undefined`.
The LLVM linker, `lld`, also accepts `--no-undefined`.

The suspect code is conditional on NOT GNU, and IS Clang, and NOT MSVC, and CMake version 3.13 or greater, and NOT WIN32. Yikes.

The suspect code was added by PR #2953 to fix issue #2951 back in May of this year. This code added the condition, is WIN32. The reporter in issue #2951 was using the LLVM linker, `lld`, as shown in his log. Their environment is Clang, MinGW, and, presumably, WIN32, which we can infer from the additional condition that was added in the PR. Interestingly, MINGW was added in CMake version 3.2. The reporter was using CMake Version 3.22.0 and should have been able to use this condition. It's not clear why this didn't work.

Regardless, I believe the root cause of this issue is the code that is intended for the Apple linker and should be guarded as such. From the man page for the OSX [linker](https://www.unix.com/man-page/OSX/1/ld/):
     `-undefined treatment`
     `Specifies how undefined symbols are to be treated. Options are: error, warning, suppress, or dynamic_lookup.  The default is error.`
